src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2088,17 +2092,65 @@ private static void UpdateFormBindingArgumentExpressions(RequestDelegateFactoryC
             Expression.NotEqual(formBuffer, Expression.Constant(null)),
             returnBufferExpr);
 
-        return Expression.Block(
-            new[] { formArgument, formReader, formDict, formBuffer },
-            Expression.TryFinally(
+        var parameterTypeNameConstant = Expression.Constant(TypeNameHelper.GetTypeDisplayName(parameter.ParameterType, fullName: false));
+        var parameterNameConstant = Expression.Constant(parameter.Name);
+
+        // try
+        // {
+        //   ProcessForm(context.Request.Form, form_dict, form_buffer);
+        //   name_reader = new FormDataReader(form_dict, CultureInfo.InvariantCulture, form_buffer.AsMemory(0, FormDataMapperOptions.MaxKeyBufferSize));
+        //   name_reader.MaxRecursionDepth = formDataMapperOptions.MaxRecursionDepth;
+        //   name_local = FormDataMapper.Map<string>(name_reader, FormDataMapperOptions);
+        // }
+        // catch (FormDataMappingException e)
+        // {
+        //    wasParamCheckFailure = true;
+        // }
+        // finally
+        // {
+        //   if (form_buffer != null)
+        //   {
+        //     ArrayPool<char>.Shared.Return(form_buffer, false);
+        //   }
+        // }
+        var bindAndCheckForm = Expression.Block(
+            new[] { formReader, formDict, formBuffer, formDataMappingException },
+            Expression.TryCatchFinally(
                 Expression.Block(
+                    typeof(void),
                     processFormExpr,
                     initializeReaderExpr,
                     setMaxRecursionDepthExpr,
                     Expression.Assign(formArgument, invokeMapMethodExpr)),
-                conditionalReturnBufferExpr),
-            formArgument
+                conditionalReturnBufferExpr,
+                Expression.Catch(formDataMappingException, Expression.Block(
+                    typeof(void),
+                    Expression.Assign(WasParamCheckFailureExpr, Expression.Constant(true)),
+                    Expression.Call(
+                        LogFormMappingFailedMethod,
+                        HttpContextExpr,
+                        parameterTypeNameConstant,
+                        parameterNameConstant,
+                        formDataMappingException,
+                        Expression.Constant(factoryContext.ThrowOnBadRequest))
+                )))
         );
+
+        // Update the existing expressions if they already exist and we're
+        // recomputing the form binding expressions after reading endpoint
+        // metadata.
+        if (index is {} i)
+        {
+            factoryContext.ParamCheckExpressions[i] = bindAndCheckForm;
```

Comment:
Looks like there might be an incorrect assumption that `ParamCheckExpressions` has items in it by the time it gets here.

Reply:
Yeah, glad the tests found this bug.

We don't populate `ParamCheckExpressions` for certain parameter types, like `HttpContext`, which means that the index can be off from the argument expression.

I changed this to only set param expressions when we apply the update after parameters has been inferred.